### ISSUE TITLE
Metric Calculation Error

### ISSUE THREAD

**Issue by enzo002021** [fastai/fastai#3537]
@jph00 
I believe that fastai calculates RMSE metric wrong as discussed on kaggle forums.  Chris Deotte (Kaggle Grandmaster) explains this issue in this thread https://www.kaggle.com/c/petfinder-pawpularity-score/discussion/293378

Key explanation in discussion:

+++++++

Since each fold is RMSE (root mean squared error), you cannot average square roots and get a square root.

If five folds (of same size) have RSME scores A, B, C, D, E respectively. Then the overall RMSE CV is sqrt(mean(A**2, B**2, C**2, D**2, E**2)). Or better yet, save each fold OOF predictions into the correct rows of a single dataframe. And then afterward, compute the RMSE on the entire dataframe.

And this isn't the only pitfall with CVs. Many public notebooks make this mistake when they compute the RMSE of a single fold. For example, I have observed that fastai computes the RMSE wrong for each fold. I think it takes the averages of all the batches, but that is incorrect (as shown above).

For example, If one epoch is 5 batches (of course there will be more batches than this) and the batches have RMSE scores A, B, C, D, E. Then the overall epoch RMSE is sqrt(mean(A**2, B**2, C**2, D**2, E**2)). The epoch does not have RSME score mean(A, B, C, D, E) which is what I think fastai does.

(This is why if you change the public FastAI notebook to batch size 2, it achieves CV RMSE score 15.0! because fastai is computing fold RMSE incorrectly).

In conclusion, to compute CV (if you can't fix the above mistakes), it is best to use your saved models after training and infer each fold yourself and save them into a single dataframe. Then compute the RMSE on this dataframe.

**Comment by warner-benjamin** [comment]
I investigated this and the fastai RSME appears to be working correctly.

The sklearn docs example below returns 0.6124
```python
from sklearn.metrics import mean_squared_error
y_true = [3, -0.5, 2, 7]
y_pred = [2.5, 0.0, 2, 8]
mean_squared_error(y_true, y_pred, squared=False)
```

The fastai metrics test for RSME also returns 0.6124
```python
from fastai.learner import Learner
from fastai.metrics import rmse
import torch

class TstLearner(Learner):
    def __init__(self,dls=None,model=None,**kwargs): self.pred,self.xb,self.yb = None,None,None

def compute_val(met, x1, x2):
    met.reset()
    vals = [0,6,15,20]
    learn = TstLearner()
    for i in range(3):
        learn.pred,learn.yb = x1[vals[i]:vals[i+1]],(x2[vals[i]:vals[i+1]],)
        met.accumulate(learn)
    return met.value
    
y_true = torch.tensor([3, -0.5, 2, 7])
y_pred = torch.tensor([2.5, 0.0, 2, 8])
compute_val(rmse, y_pred, y_true)
```

And if we construct a simple tabular training loop, the metric also returns 0.6124 for all batch sizes [1,4]:
```python
from fastai.tabular.all import *

# add zeros to be the training set
df = pd.DataFrame.from_dict({'preds': [0, 2.5, 0.0, 2, 8], 'true': [0, 3, -0.5, 2, 7]})

# create the datablock with 0s in training set, rest in valid set
tp = TabularPandas(df, cont_names='preds', y_names=['true'], splits=([0], [i for i in range(1,5)]))

# uncomment to verify the valid dataloader has all the rsme test values
# dls = tp.dataloaders(bs=2)
# dls.valid_ds

# dummy model which just passes input, __init__ is so the optimizer doesn't error out
class DummyModel(Module):
    def __init__(self): self.a = nn.Parameter(torch.randn(1))

    def forward(self, *x):
        return x[1]

# no matter the batch size (1,2,3,4), we expect the metric result to be #0.6124. Which it is
rmse_metrics = []
for bs in range(1,5):
    dls = tp.dataloaders(bs=bs)
    learn = Learner(dls, DummyModel(), loss_func=MSELossFlat(), metrics=rmse)
    rmse_metrics.append(learn.validate()[1])

print(rmse_metrics)
```

**Comment by jph00** [comment]
Many thanks for your clear analysis @warner-benjamin
